jumanji/environments/logic/sliding_tile_puzzle/env.py

```diff
@@ -0,0 +1,285 @@
+        )
+        # Check if the puzzle is solved
+        done = jnp.array_equal(updated_puzzle, self.solved_puzzle)
+        horizon_reached = state.step_count >= self.time_limit
+
+        # Update the action mask
+        action_mask = self._get_valid_actions(updated_empty_tile_position)
+
+        next_state = State(
+            puzzle=updated_puzzle,
+            empty_tile_position=updated_empty_tile_position,
+            key=state.key,
+            step_count=state.step_count + 1,
+        )
+        obs = Observation(
+            puzzle=updated_puzzle,
+            empty_tile_position=updated_empty_tile_position,
+            action_mask=action_mask,
+        )
+
+        reward = self.reward_fn(state, action, next_state, self.solved_puzzle)
+        extras = self._get_extras(next_state)
+
+        timestep = jax.lax.cond(
+            done | (next_state.step_count >= self.time_limit),
+            lambda: termination(
+                reward=reward,
+                observation=obs,
+                extras=extras,
+            ),
+            lambda: transition(
+                reward=reward,
+                observation=obs,
+                extras=extras,
+            ),
+        )
+
+        return next_state, timestep
+
+    def _move_empty_tile(
+        self,
+        puzzle: chex.Array,
+        empty_tile_position: chex.Array,
+        action: chex.Array,
+    ) -> chex.Array:
+        """Moves the empty tile in the given direction and returns the updated puzzle and reward."""
+
+        # Compute the new position
+        new_empty_tile_position = empty_tile_position + MOVES[action]
+
+        # Predicate for the conditional
+        is_valid_move = jnp.all(
+            (new_empty_tile_position >= 0)
+            & (new_empty_tile_position < self.generator.grid_size)
+        )
+
+        def valid_move(puzzle: chex.Array) -> Tuple[chex.Array, Tuple[int, int]]:
+            # Swap the empty tile and the tile at the new position
+            updated_puzzle = puzzle.at[tuple(empty_tile_position)].set(
+                puzzle[tuple(new_empty_tile_position)]
+            )
+            updated_puzzle = updated_puzzle.at[tuple(new_empty_tile_position)].set(0)
+
+            return updated_puzzle, new_empty_tile_position
+
+        def invalid_move(puzzle: chex.Array) -> Tuple[chex.Array, Tuple[int, int]]:
+            # If the move is not valid, return the original puzzle
+            return puzzle, empty_tile_position
+
+        return lax.cond(is_valid_move, valid_move, invalid_move, puzzle)
+
+    def _get_valid_actions(self, empty_tile_position: chex.Array) -> chex.Array:
+        # Compute the new positions if these movements are applied
+        new_positions = empty_tile_position + MOVES
+
+        # Check if the new positions are within the grid boundaries
+        valid_moves_mask = jnp.all(
+            (new_positions >= 0) & (new_positions < self.generator.grid_size), axis=-1
+        )
```

Comment:
I am not sure I understand the `all(..., axis=-1)` here. We want to compare boundaries for all actions, so no `all` for me.

Reply:
all is on the coordinate level, all good